Refactor the notification options into a component

JayyajGH opened the pull request:
This is a rethink of the notification options component.

Generally I'm happy with the component now although I need to give it some more testing hence the draft PR.  The only point of contention is the about me modal which I've been struggling with a bit.

So at the moment the component can only have a single element due to the way the parent is set up.  The parent in some cases is a UL and so the component can't have a wrapper DIV or it breaks the structure.

Thinking about ownership, it doesn't seem right for the about me modal to live in the notification component.  It has got nothing to do with notifications.  It just happens to be called from that component.  This modal is called from other places and these places currently also define a version of the modal. (Potential future refactor...)

This PR leaves the about me modal in the default layout and just raises an event from the notifications component to show it.  To me this feels much nicer as it keeps the notification component clean and focused on one thing.

Thoughts?

edwh commented:
That sounds fine.

In general we could have a bunch of modals which get triggered from 
different places.  The microvolunteering stuff I'm working on is a bit 
like that.  So if you wanted to feel better about it, you could create a 
wrapper component for that and listen on $root as an event bus.

------ Original Message ------
From: "JayyajGH" <notifications@github.com>
To: "Freegle/iznik-nuxt" <iznik-nuxt@noreply.github.com>
Cc: "Subscribed" <subscribed@noreply.github.com>
Sent: 16/11/2020 12:00:59
Subject: [Freegle/iznik-nuxt] Refactor the notification options into a 
component (#334)

>This is a rethink of the notification options component.
>
>Generally I'm happy with the component now although I need to give it 
>some more testing hence the draft PR. The only point of contention is 
>the about me modal which I've been struggling with a bit.
>
>So at the moment the component can only have a single element due to 
>the way the parent is set up. The parent in some cases is a UL and so 
>the component can't have a wrapper DIV or it breaks the structure.
>
>Thinking about ownership, it doesn't seem right for the about me modal 
>to live in the notification component. It has got nothing to do with 
>notifications. It just happens to be called from that component. This 
>modal is called from other places and these places currently also 
>define a version of the modal. (Potential future refactor...)
>
>This PR leaves the about me modal in the default layout and just raises 
>an event from the notifications component to show it. To me this feels 
>much nicer as it keeps the notification component clean and focused on 
>one thing.
>
>Thoughts?
>
>--------------------------------------------------------------------------------
>You can view, comment on, or merge this pull request online at:
>   https://github.com/Freegle/iznik-nuxt/pull/334
>
>Commit Summary
>Refactor the notification options into a component
>File Changes
>Acomponents/NotificationOptions.vue 
><https://github.com/Freegle/iznik-nuxt/pull/334/files#diff-65a3c422c60610e95e2b61dbc57c4b8832f7c51b82db45fe66bfea80f513d7ed> 
>(182)
>Mlayouts/default.vue 
><https://github.com/Freegle/iznik-nuxt/pull/334/files#diff-433a9abd4ca70520f69fc064c160bb093918ce26dda6087a8bbfdbe2895d1944> 
>(144)
>Patch Links:
>https://github.com/Freegle/iznik-nuxt/pull/334.patch
>https://github.com/Freegle/iznik-nuxt/pull/334.diff
>—
>You are receiving this because you are subscribed to this thread.
>Reply to this email directly, view it on GitHub 
><https://github.com/Freegle/iznik-nuxt/pull/334>, or unsubscribe 
><https://github.com/notifications/unsubscribe-auth/AAAN26FT33SGWIG2POKCV7TSQEH7XANCNFSM4TXCKKMQ>.
>

JayyajGH commented:
I'll leave this as it is for now and look into the modals as a phase 2 I think.  I'll give this a proper test and then un-draft it.

JayyajGH commented:
This is now ready for review.  I have tried to leave most of the code as it was e.g. the existing JS and CSS. There is some more refactoring in those areas but that can be done as another piece of work.

edwh commented:
Thanks - I'll have a look today.

------ Original Message ------
From: "JayyajGH" <notifications@github.com>
To: "Freegle/iznik-nuxt" <iznik-nuxt@noreply.github.com>
Cc: "Edward Hibbert" <edward@ehibbert.org.uk>; "Comment" 
<comment@noreply.github.com>
Sent: 17/11/2020 09:15:50
Subject: Re: [Freegle/iznik-nuxt] Refactor the notification options into 
a component (#334)

>
>This is now ready for review. I have tried to leave most of the code as 
>it was e.g. the existing JS and CSS. There is some more refactoring in 
>those areas but that can be done as another piece of work.
>
>—
>You are receiving this because you commented.
>Reply to this email directly, view it on GitHub 
><https://github.com/Freegle/iznik-nuxt/pull/334#issuecomment-728796247>, 
>or unsubscribe 
><https://github.com/notifications/unsubscribe-auth/AAAN26HJ3WEQGEXXHHXZ44DSQI5MNANCNFSM4TXCKKMQ>.
>

JayyajGH commented:
No rush.  It's part of a long journey to tidy default.vue...

edwh commented:
Can I dump something into that?

Someone pointed out that on pages with the sidebars, you get the 
"About/Terms" etc links.  But on pages where we don't have these (for 
clutter reasons) you don't.  That's probably wrong.  Those links are 
really site wide so they should be in the layout, I think.

------ Original Message ------
From: "JayyajGH" <notifications@github.com>
To: "Freegle/iznik-nuxt" <iznik-nuxt@noreply.github.com>
Cc: "Edward Hibbert" <edward@ehibbert.org.uk>; "Comment" 
<comment@noreply.github.com>
Sent: 17/11/2020 09:44:33
Subject: Re: [Freegle/iznik-nuxt] Refactor the notification options into 
a component (#334)

>
>No rush. It's part of a long journey to tidy default.vue...
>
>—
>You are receiving this because you commented.
>Reply to this email directly, view it on GitHub 
><https://github.com/Freegle/iznik-nuxt/pull/334#issuecomment-728812378>, 
>or unsubscribe 
><https://github.com/notifications/unsubscribe-auth/AAAN26CW4ONYHLN3DEUCUHTSQJAYDANCNFSM4TXCKKMQ>.
>

JayyajGH commented:
So every page should have the about/terms displayed on the sidebar?  We could just add them in to the missing pages as a quick fix or just wait until the layout pages are refactored.

edwh commented:
Yes.

It's not urgent, so I'd rather it was part of the layout, assuming you 
agree that's where it should be.

------ Original Message ------
From: "JayyajGH" <notifications@github.com>
To: "Freegle/iznik-nuxt" <iznik-nuxt@noreply.github.com>
Cc: "Edward Hibbert" <edward@ehibbert.org.uk>; "Comment" 
<comment@noreply.github.com>
Sent: 17/11/2020 09:53:36
Subject: Re: [Freegle/iznik-nuxt] Refactor the notification options into 
a component (#334)

>
>So every page should have the about/terms displayed on the sidebar? We 
>could just add them in to the missing pages as a quick fix or just wait 
>until the layout pages are refactored.
>
>—
>You are receiving this because you commented.
>Reply to this email directly, view it on GitHub 
><https://github.com/Freegle/iznik-nuxt/pull/334#issuecomment-728817545>, 
>or unsubscribe 
><https://github.com/notifications/unsubscribe-auth/AAAN26C2B3SXM76UIQTOMKLSQJB2BANCNFSM4TXCKKMQ>.
>

JayyajGH commented:
0k cool.  Do they need to be so obvious?  Just looking at other infinite scroll websites (FB and twitter) and they seem to hide their terms etc. away in menus.  Is that an option?  Especially as we already hide them on the simple view

edwh commented:
They're hidden on smaller screens, too, of course.

We have links in the Help section, which you'd have thought would be 
enough.  But we do get a steady stream of people who are asking for and 
cannot find them.  For example we recently had someone who wanted to 
donate, couldn't find the donate link, googled "freegle donate", and 
then contacted us to say it was hard to find.

So in the cases where the sidebar is visible but empty, I'd prefer it if 
they were there.

------ Original Message ------
From: "JayyajGH" <notifications@github.com>
To: "Freegle/iznik-nuxt" <iznik-nuxt@noreply.github.com>
Cc: "Edward Hibbert" <edward@ehibbert.org.uk>; "Comment" 
<comment@noreply.github.com>
Sent: 17/11/2020 10:02:52
Subject: Re: [Freegle/iznik-nuxt] Refactor the notification options into 
a component (#334)

>
>0k cool. Do they need to be so obvious? Just looking at other infinite 
>scroll websites (FB and twitter) and they seem to hide their terms etc. 
>away in menus. Is that an option? Especially as we already hide them on 
>the simple view
>
>—
>You are receiving this because you commented.
>Reply to this email directly, view it on GitHub 
><https://github.com/Freegle/iznik-nuxt/pull/334#issuecomment-728823173>, 
>or unsubscribe 
><https://github.com/notifications/unsubscribe-auth/AAAN26CDNAM7TEGD5G5BI7DSQJC4ZANCNFSM4TXCKKMQ>.
>

JayyajGH commented:
In that case do we need all of them?  Can we just have donate on the main screen and hide the others?

edwh commented:
We get questions about the others too.

------ Original Message ------
From: "JayyajGH" <notifications@github.com>
To: "Freegle/iznik-nuxt" <iznik-nuxt@noreply.github.com>
Cc: "Edward Hibbert" <edward@ehibbert.org.uk>; "Comment" 
<comment@noreply.github.com>
Sent: 17/11/2020 10:08:40
Subject: Re: [Freegle/iznik-nuxt] Refactor the notification options into 
a component (#334)

>
>In that case do we need all of them? Can we just have donate on the 
>main screen and hide the others?
>
>—
>You are receiving this because you commented.
>Reply to this email directly, view it on GitHub 
><https://github.com/Freegle/iznik-nuxt/pull/334#issuecomment-728826628>, 
>or unsubscribe 
><https://github.com/notifications/unsubscribe-auth/AAAN26CNWYZ472SCFTRHXPLSQJDSRANCNFSM4TXCKKMQ>.
>

JayyajGH commented:
Thought I'd check! :-)

JayyajGH commented:
I've now got the chat stuff pulled out into a component as well.  Do you want to deal with that separately or with this?

edwh commented:
With this - not got to it yet.

------ Original Message ------
From: "JayyajGH" <notifications@github.com>
To: "Freegle/iznik-nuxt" <iznik-nuxt@noreply.github.com>
Cc: "Edward Hibbert" <edward@ehibbert.org.uk>; "Comment" 
<comment@noreply.github.com>
Sent: 17/11/2020 14:52:55
Subject: Re: [Freegle/iznik-nuxt] Refactor the notification options into 
a component (#334)

>
>I've now got the chat stuff pulled out into a component as well. Do you 
>want to deal with that separately or with this?
>
>—
>You are receiving this because you commented.
>Reply to this email directly, view it on GitHub 
><https://github.com/Freegle/iznik-nuxt/pull/334#issuecomment-728979986>, 
>or unsubscribe 
><https://github.com/notifications/unsubscribe-auth/AAAN26EXEAYG7LX6NMIIQK3SQKE4PANCNFSM4TXCKKMQ>.
>

JayyajGH commented:
Was there any reason you can remember for adding a "to" attribute on the chat anchor?  It's all run off JS and there is preventdefault set so I can't really see a use for it.  I've just added a href="#" to both small and large screens. (Although technically it isn't required but bootstrap adds it to b-nav-item so I've added it for consistency)

edwh commented:
Where precisely?

------ Original Message ------
From: "JayyajGH" <notifications@github.com>
To: "Freegle/iznik-nuxt" <iznik-nuxt@noreply.github.com>
Cc: "Edward Hibbert" <edward@ehibbert.org.uk>; "Comment" 
<comment@noreply.github.com>
Sent: 17/11/2020 15:04:41
Subject: Re: [Freegle/iznik-nuxt] Refactor the notification options into 
a component (#334)

>
>Was there any reason you can remember for adding a "to" attribute on 
>the chat anchor? It's all run off JS and there is preventdefault set so 
>I can't really see a use for it. I've just added a href="#" to both 
>small and large screens. (Although technically it isn't required but 
>bootstrap adds it to b-nav-item so I've added it for consistency)
>
>—
>You are receiving this because you commented.
>Reply to this email directly, view it on GitHub 
><https://github.com/Freegle/iznik-nuxt/pull/334#issuecomment-728988033>, 
>or unsubscribe 
><https://github.com/notifications/unsubscribe-auth/AAAN26H2J6RDNTSTE67QVT3SQKGITANCNFSM4TXCKKMQ>.
>

JayyajGH commented:
In default.vue there is a b-nav-item for chats with an id of "menu-option-chat"

edwh commented:
I think that's a hangover from earlier code.

------ Original Message ------
From: "JayyajGH" <notifications@github.com>
To: "Freegle/iznik-nuxt" <iznik-nuxt@noreply.github.com>
Cc: "Edward Hibbert" <edward@ehibbert.org.uk>; "Comment" 
<comment@noreply.github.com>
Sent: 17/11/2020 15:07:39
Subject: Re: [Freegle/iznik-nuxt] Refactor the notification options into 
a component (#334)

>
>In default.vue there is a b-nav-item for chats with an id of 
>"menu-option-chat"
>
>—
>You are receiving this because you commented.
>Reply to this email directly, view it on GitHub 
><https://github.com/Freegle/iznik-nuxt/pull/334#issuecomment-728989851>, 
>or unsubscribe 
><https://github.com/notifications/unsubscribe-auth/AAAN26DZ26Z2KMMDGVIQFNDSQKGTXANCNFSM4TXCKKMQ>.
>

JayyajGH commented:
OK cool.  It works ok without it. :-)

Although just noticed a bit of an accessibility issue with the mobile menu so I'll fix that whilst I'm at it

JayyajGH commented:
I've just pushed the chats refactor too.  I haven't given this a great test yet though.  It's had minimal testing.

edwh commented:
Ok.

It's good that you're doing this kind of stuff.  It would never reach 
the top of my list, but it stops code decay.

------ Original Message ------
From: "JayyajGH" <notifications@github.com>
To: "Freegle/iznik-nuxt" <iznik-nuxt@noreply.github.com>
Cc: "Edward Hibbert" <edward@ehibbert.org.uk>; "Comment" 
<comment@noreply.github.com>
Sent: 17/11/2020 15:18:22
Subject: Re: [Freegle/iznik-nuxt] Refactor the notification options into 
a component (#334)

>
>I've just pushed the chats refactor too. I haven't given this a great 
>test yet though. It's had minimal testing.
>
>—
>You are receiving this because you commented.
>Reply to this email directly, view it on GitHub 
><https://github.com/Freegle/iznik-nuxt/pull/334#issuecomment-728997033>, 
>or unsubscribe 
><https://github.com/notifications/unsubscribe-auth/AAAN26FL5MZJ742WS657MQLSQKH35ANCNFSM4TXCKKMQ>.
>

JayyajGH commented:
It's like how the Vue team used to work.  The core team (you and Chris) work on new features and the community (me) work on bugs and smaller improvements. :-)

edwh commented on components/NotificationOptions.vue:
```diff
@@ -0,0 +1,185 @@
+  },
+  computed: {
+    notificationType() {
+      return this.smallScreen ? 'b-dropdown' : 'b-nav-item-dropdown'
```
Can you comment saying why we do this?

JayyajGH replied:
Do you mean technically as in how the code is used or why we have two different styles of menu?

JayyajGH replied:
Would something like this make sense?

// A different component needs to be created depending on the context in which it's used

edwh commented on components/ChatMenu.vue:
```diff
@@ -0,0 +1,76 @@
+  },
+  watch: {
+    chatCount: function() {
+      this.$emit('chat-count', this.chatCount)
```
Wouldn't it be better to emit update:chat-count so that the parent can use .sync?

JayyajGH replied:
I avoided using sync as it's been deprecated in v3 hasn't it?

edwh commented on components/NotificationOptions.vue:
```diff
@@ -0,0 +1,185 @@
+  },
+  watch: {
+    unreadNotificationCount: function() {
+      this.$emit('unread-notification-count', this.unreadNotificationCount)
```
Emit update:?

edwh commented on components/NotificationOptions.vue:
```diff
@@ -0,0 +1,185 @@
+        this.busy = true
+        this.$store
+          .dispatch('notifications/fetchNextListChunk')
+          .then(() => {
```
When I wrote this I was clueless about await.  While we're restructuring this code, could we rework it to use that?

JayyajGH replied:
Just checked in some code to use await.  It looks much tidier.  You probably should take a proper look at this bit.  I think I've got everything but was unsure about some of the infinite-loading values.  e.g. Do we still need the busy property?

edwh commented on layouts/default.vue:
```diff
@@ -392,9 +307,8 @@ export default {
       this.startNCHAN(me.id)
 
-      // Get notifications and chats and poll regularly for new ones.  Would be nice if this was event driven instead but requires server work.
+      // Get chats and poll regularly for new ones.  Would be nice if this was event driven instead but requires server work.
```
This comment suggests a timer which has now moved into the store, in notifications.js.  Can you move the comment there?

edwh commented on components/ChatMenu.vue:
```diff
@@ -0,0 +1,76 @@
+  },
+  computed: {
+    chatType() {
+      return this.smallScreen ? 'a' : 'b-nav-item'
```
Can you comment why we do this?

edwh commented on layouts/default.vue:
```diff
@@ -392,9 +307,8 @@ export default {
 
-      // Get notifications and chats and poll regularly for new ones.  Would be nice if this was event driven instead but requires server work.
+      // Get chats and poll regularly for new ones.  Would be nice if this was event driven instead but requires server work.
       this.fetchLatestChats()
```
We've lost a call to updateNotifications here. Is that intentional?

JayyajGH replied:
Which one?  The updateNotifications call which was in mounted in default.vue has been moved into the mounted of notificationoptions.vue.

Although I have noticed that I can also move the fetchchats in to the relevant component....

edwh reviewed (commented):
This rips a good chunk of code out, which is nice.  A few comments.

edwh commented:
.sync itself is going, but I think you still emit the update event.  
That's what 
https://v3.vuejs.org/guide/migration/v-model.html#using-v-bind-sync 
suggests to me.

------ Original Message ------
From: "JayyajGH" <notifications@github.com>
To: "Freegle/iznik-nuxt" <iznik-nuxt@noreply.github.com>
Cc: "Edward Hibbert" <edward@ehibbert.org.uk>; "Comment" 
<comment@noreply.github.com>
Sent: 18/11/2020 09:08:16
Subject: Re: [Freegle/iznik-nuxt] Refactor the notification options into 
a component (#334)

>
>@JayyajGH commented on this pull request.
>
>--------------------------------------------------------------------------------
>In components/ChatMenu.vue 
><https://github.com/Freegle/iznik-nuxt/pull/334#discussion_r525922131>:
>
> > +      required: false,
>+      default: false
>+    }
>+  },
>+  computed: {
>+    chatType() {
>+      return this.smallScreen ? 'a' : 'b-nav-item'
>+    },
>+    chatCount() {
>+      // Don't show so many that the layout breaks.
>+      return Math.min(99, this.$store.getters['chats/unseenCount'])
>+    }
>+  },
>+  watch: {
>+    chatCount: function() {
>+      this.$emit('chat-count', this.chatCount)
>
>I avoided using sync as it's been deprecated in v3 hasn't it?
>
>—
>You are receiving this because you commented.
>Reply to this email directly, view it on GitHub 
><https://github.com/Freegle/iznik-nuxt/pull/334#discussion_r525922131>, 
>or unsubscribe 
><https://github.com/notifications/unsubscribe-auth/AAAN26CO5WHO3B2PH3KFX7TSQOFIBANCNFSM4TXCKKMQ>.
>

JayyajGH commented:
Ah so it's actually quite similar.  I obviously hadn't read it properly!

edwh commented:
In both cases, why we're using different components.  It's just not 
locally obvious from looking at that component, and because it's unusual 
to do this I think it bears a bit more commenting.

------ Original Message ------
From: "JayyajGH" <notifications@github.com>
To: "Freegle/iznik-nuxt" <iznik-nuxt@noreply.github.com>
Cc: "Edward Hibbert" <edward@ehibbert.org.uk>; "Comment" 
<comment@noreply.github.com>
Sent: 18/11/2020 09:05:29
Subject: Re: [Freegle/iznik-nuxt] Refactor the notification options into 
a component (#334)

>
>@JayyajGH commented on this pull request.
>
>--------------------------------------------------------------------------------
>In components/NotificationOptions.vue 
><https://github.com/Freegle/iznik-nuxt/pull/334#discussion_r525920379>:
>
> > +    Notification
>+  },
>+  props: {
>+    distance: {
>+      type: Number,
>+      required: true
>+    },
>+    smallScreen: {
>+      type: Boolean,
>+      required: false,
>+      default: false
>+    }
>+  },
>+  computed: {
>+    notificationType() {
>+      return this.smallScreen ? 'b-dropdown' : 'b-nav-item-dropdown'
>
>Do you mean technically as in how the code is used or why we have two 
>different styles of menu?
>
>—
>You are receiving this because you commented.
>Reply to this email directly, view it on GitHub 
><https://github.com/Freegle/iznik-nuxt/pull/334#discussion_r525920379>, 
>or unsubscribe 
><https://github.com/notifications/unsubscribe-auth/AAAN26FZSM2WPY5MCW64OB3SQOE5TANCNFSM4TXCKKMQ>.
>

JayyajGH commented:
I'm actually still a bit confused as to why we are using different component!  I was going to start looking at that once I could see the wood for the trees in default.vue.  In chats for example, small screens use b-nav-item which creates an anchor and large screens directly use an anchor.

JayyajGH commented:
Looks like we could pull fetchLatestChats out into the store in the same way as notifications works.

edwh commented:
Catching up:

- I can't remember why we use different components either, which is what makes me think it either needs a comment or might be wrong and then could be removed.
- For chats, I think there is more refactoring that is needed, but it's not clear what.  Maybe move chat updating out of pages into the store, as you say, but there will be subtleties about when we need to ask for updates.  Definitely not part of this - you're tidying up the default vue here.

So this might be nearly there?

JayyajGH commented:
For now I've added the comment:

// A different component needs to be created depending on the context in which it's used

Does that sort of cover it for now?  Once default.vue is stripped down to it's basics then we can start looking into how the menu itself works and maybe working out why it is how it is.  Moving it out into a component isn't making it any worse than it was before or any more confusing.  If anything it's actually making it more obvious.

JayyajGH commented:
Chats - My new component is basically for the menu and that's it.  I think the next step will be to move the chat update stuff out into the store maybe but I can look at that as a separate piece of work.

JayyajGH commented:
You did mention earlier "We've lost a call to updateNotifications here. Is that intentional?"  I think they've all been included.  Could you double check that please.  I'm hoping you're wrong! :-)

JayyajGH commented:
And also could you take a closer look at the promise/await refactor I did please.  Mainly to check that I got the vue-infinite stuff correct rather than the promises.

edwh commented:
I've merged master to make it easier to compare.  If I do a compare against master I see this:

![image](https://user-images.githubusercontent.com/56696/99654251-f40ad400-2a51-11eb-9476-3993a96f5605.png)

JayyajGH commented:
Ah yes.  I've moved that into the mounted method in the component.

edwh commented:
So you have.

The await stuff looks fine.  The busy field is not needed if it ever was, so you could lose that.  Then I think it's good to go.

JayyajGH commented:
In that case we're good to go!

edwh commented:
Phew!

JayyajGH commented:
:-)  That was quite painful.  At least we got there in the end.  default.vue is looking much more manageable already
